optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/bi/BavetScoringBiConstraintStream.java

```diff
@@ -111,68 +108,37 @@ public boolean guaranteesDistinct() {
-                scoreImpacter = (A a, B b, Consumer<Score<?>> matchScoreConsumer) -> {
-                    BigDecimal matchWeight = bigDecimalMatchWeigher.apply(a, b);
-                    constraint.assertCorrectImpact(matchWeight);
-                    return castedWeightedScoreImpacter.impactScore(matchWeight, matchScoreConsumer);
-                };
-            } else if (noMatchWeigher) {
-                scoreImpacter = (A a, B b, Consumer<Score<?>> matchScoreConsumer) -> castedWeightedScoreImpacter
-                        .impactScore(BigDecimal.ONE, matchScoreConsumer);
-            } else {
-                throw new IllegalStateException("The matchWeigher of " + BiConstraintStream.class.getSimpleName()
-                        + ".penalize(matchWeigher) of the constraint (" + constraint.getConstraintId()
-                        + ") must return a " + BigDecimal.class.getSimpleName() + ".\n" +
-                        "Maybe switch to penalizeBigDecimal() or rewardBigDecimal().");
-            }
+        ScoreInliner<?> scoreInliner = buildPolicy.getSession().getScoreInliner();
+        WeightedScoreImpacter weightedScoreImpacter = scoreInliner.buildWeightedScoreImpacter(constraint);
+        BiFunction<A, B, UndoScoreImpacter> scoreImpacter;
+        if (intMatchWeigher != null) {
+            scoreImpacter = (a, b) -> {
+                int matchWeight = intMatchWeigher.applyAsInt(a, b);
+                constraint.assertCorrectImpact(matchWeight);
+                return weightedScoreImpacter.impactScore(matchWeight, () -> asList(a, b));
+            };
+        } else if (longMatchWeigher != null) {
+            scoreImpacter = (a, b) -> {
+                long matchWeight = longMatchWeigher.applyAsLong(a, b);
+                constraint.assertCorrectImpact(matchWeight);
+                return weightedScoreImpacter.impactScore(matchWeight, () -> asList(a, b));
+            };
+        } else if (bigDecimalMatchWeigher != null) {
+            scoreImpacter = (a, b) -> {
+                BigDecimal matchWeight = bigDecimalMatchWeigher.apply(a, b);
+                constraint.assertCorrectImpact(matchWeight);
+                return weightedScoreImpacter.impactScore(matchWeight, () -> asList(a, b));
+            };
+        } else if (noMatchWeigher) {
+            scoreImpacter = (a, b) -> weightedScoreImpacter.impactScore(1, () -> asList(a, b));
         } else {
-            throw new IllegalStateException("Unsupported weightedScoreImpacter (" + weightedScoreImpacter + ").");
+            throw new IllegalStateException("The matchWeigher of " + BiConstraintStream.class.getSimpleName()
+                    + ".penalize(matchWeigher) of the constraint (" + constraint.getConstraintId()
+                    + ") must return " + weightedScoreImpacter.getExpectedMatchWeightType().getSimpleName() + ".\n" +
+                    "Maybe switch to another type-specific overload of penalize() or reward().");
         }
```

Comment:
See the simplification I was able to make because of the changes to `WeightedScoreImpacter` I pointed out above.